In [15]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

file_name = "/content/uy.csv"

df = pd.read_csv(file_name)

df.head(10)

# df.describe()

# df.info()

,narx,maydon,yotoqxona,hammom,qavatlar
0,1.187601e+06,1520,5,2,1
1,2.150566e+06,4850,5,2,1
2,2.312066e+06,3980,4,2,1
3,1.912090e+06,3200,2,3,1
4,1.959511e+05,1560,1,1,2
5,4.321134e+05,1210,2,1,2
6,1.192251e+06,2380,1,2,3
7,4.091128e+05,700,4,1,1
8,6.800597e+05,1520,4,1,3
9,7.618992e+05,1710,5,2,2


In [6]:
df_simpl = df[['maydon', 'narx']].copy()

# Bo'sh qiymatlar bo'lsa, tashlaymiz
df_simpl.dropna(inplace=True)

# X va Y ni qiymatlarni ajratamiz
X = df_simpl['maydon'].values  # Mustaqil o'zgaruvchi (maydon)
y = df_simpl['narx'].values    # Mustaqil bo'lmagan o'zgaruvchi (narx)

In [20]:
print(type(X))
X.shape

<class 'numpy.ndarray'>


(1000,)

In [7]:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=55)
print(x_train.shape, x_test.shape)
print(y_train.shape, y_test.shape)

(800,) (200,)
(800,) (200,)


In [27]:
import numpy as np

def my_train_test_split(X, y, test_size=0.2, random_state=None):

    # Random seed
    if random_state is not None:
        np.random.seed(random_state)

    # Dataset size
    n = len(X)

    # Index yaratish
    indices = np.arange(n)

    # Aralashtirish
    np.random.shuffle(indices)

    # Test dataset soni
    test_count = int(n * test_size)

    # Train/Test indexlar
    test_indices = indices[:test_count]
    train_indices = indices[test_count:]

    # Data ajratish
    x_train = X[train_indices]
    x_test = X[test_indices]

    y_train = y[train_indices]
    y_test = y[test_indices]

    return x_train, x_test, y_train, y_test

In [28]:
x_train, x_test, y_train, y_test = my_train_test_split(X, y, test_size = 0.2, random_state=55)
print(x_train.shape, x_test.shape)
print(y_train.shape, y_test.shape)

(800,) (200,)
(800,) (200,)


In [38]:
from sklearn.preprocessing import MinMaxScaler

x_scaler = MinMaxScaler()
y_scaler = MinMaxScaler()

x_train = x_scaler.fit_transform(x_train.reshape(-1,1))
x_test = x_scaler.transform(x_test.reshape(-1,1))

y_train = y_scaler.fit_transform(y_train.reshape(-1,1))
y_test = y_scaler.transform(y_test.reshape(-1,1))

In [39]:
print(x_train.shape)


(800, 1)


In [40]:
y_train = y_train.flatten()
y_test = y_test.flatten()
x_train = x_train.flatten()
x_test = x_test.flatten()



In [41]:
print(x_train.shape)

(800,)


In [42]:
# Parametrlarga boshlang'ich qiymat berish
b = 0.5  # Boshlang'ich kesish nuqtasi
m_ = 0.5 # Boshlang'ich o'zgarish qiymati
print(f"b: {b}, m: {m_}")
# O'rganish sur'ati
alpha = 0.01
print(f"alpha {alpha}")

# Iteratsiyalar soni
num_iterations = 1000

N = len(x_train)  # namunalar soni
print(f"N {N}")

for i in range(num_iterations):
    Y_hat = b + m_ * x_train  # Bashorat
    error = Y_hat - y_train   # Xatolik
    db = (1/N) * np.sum(error)  # ∂J/∂b
    dm = (1/N) * np.sum(error * x_train)  # ∂J/∂m
    b = b - alpha * db
    m_ = m_ - alpha * dm

    if i % 100 == 0:
        cost = (1/(2*N)) * np.sum(error**2)
        print(f"Iteratsiya {i}: jarima={cost:.2f}, b={b:.2f}, m={m_:.4f}")

b: 0.5, m: 0.5
alpha 0.01
N 800
Iteratsiya 0: jarima=0.03, b=0.50, m=0.4989
Iteratsiya 100: jarima=0.01, b=0.36, m=0.4425
Iteratsiya 200: jarima=0.01, b=0.32, m=0.4366
Iteratsiya 300: jarima=0.00, b=0.30, m=0.4440
Iteratsiya 400: jarima=0.00, b=0.29, m=0.4545
Iteratsiya 500: jarima=0.00, b=0.28, m=0.4653
Iteratsiya 600: jarima=0.00, b=0.28, m=0.4758
Iteratsiya 700: jarima=0.00, b=0.27, m=0.4856
Iteratsiya 800: jarima=0.00, b=0.27, m=0.4949
Iteratsiya 900: jarima=0.00, b=0.26, m=0.5036


In [43]:
y_pred_scaled = b + m_ * x_test

In [44]:
# Original qiymatga qaytarish
y_pred = y_scaler.inverse_transform(y_pred_scaled.reshape(-1,1))

y_test_original = y_scaler.inverse_transform(y_test.reshape(-1,1))

In [46]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import numpy as np

# MAE
mae = mean_absolute_error(y_test_original, y_pred)

# MSE
mse = mean_squared_error(y_test_original, y_pred)

# RMSE
rmse = np.sqrt(mse)

# R²
r2 = r2_score(y_test_original, y_pred)

print("MAE :", mae)
print("MSE :", mse)
print("RMSE:", rmse)
print("R²  :", r2)

MAE : 0.07481050845175338
MSE : 0.008460890365155646
RMSE: 0.09198309825808025
R²  : 0.7539309017404998


In [47]:
mae_scaled = mean_absolute_error(y_test, y_pred_scaled)
print(mae_scaled)

0.07481050845175338


In [48]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# DATASET
file_name = "/content/uy.csv"

df = pd.read_csv(file_name)

# KERAKLI COLUMN
df_simpl = df[['maydon', 'narx']].copy()

# NULL VALUE
df_simpl.dropna(inplace=True)

# X va y
X = df_simpl[['maydon']].values
y = df_simpl['narx'].values.reshape(-1,1)

# TRAIN / TEST
x_train, x_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=55
)

# SCALER
x_scaler = MinMaxScaler()
y_scaler = MinMaxScaler()

# SCALE
x_train = x_scaler.fit_transform(x_train)
x_test = x_scaler.transform(x_test)

y_train = y_scaler.fit_transform(y_train)
y_test = y_scaler.transform(y_test)

# MODEL
model = LinearRegression()

# TRAIN
model.fit(x_train, y_train)

# PREDICTION
y_pred = model.predict(x_test)

# MSE
mse = mean_squared_error(y_test, y_pred)

print("MSE:", mse)

print()

print("Prediction:")
print(y_pred[:5])

print()

print("Real:")
print(y_test[:5])

MSE: 0.007872511397528481

Prediction:
[[0.53727289]
 [0.24352816]
 [0.57822768]
 [0.35791913]
 [0.56834204]]

Real:
[[0.43618584]
 [0.39682336]
 [0.47621257]
 [0.29292919]
 [0.6690969 ]]


In [49]:
import numpy as np

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

# MAE
mae = mean_absolute_error(y_test, y_pred)

# MSE
mse = mean_squared_error(y_test, y_pred)

# RMSE
rmse = np.sqrt(mse)

# R²
r2 = r2_score(y_test, y_pred)

print("MAE :", mae)
print("MSE :", mse)
print("RMSE:", rmse)
print("R²  :", r2)

MAE : 0.0724330521537435
MSE : 0.007872511397528481
RMSE: 0.08872717395211278
R²  : 0.7710427984499909


Bu scratch dan qilgan natijamiz:

MAE : 0.07481050845175338

MSE : 0.008460890365155646

RMSE: 0.09198309825808025

R²  : 0.7539309017404998